# 4.5 Data Consistency Checks

## This script continas the following points:

#### 1. Import libraries
#### 2. Import data
#### 3. Follow along with practice task
#### 4. Task 4.5

# 01. Importing libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 02. Importing Data

In [2]:
# import data
path = r'D:\CareerFoundary\Immersion into Data Analystics\4 Python Fundamentals for Data Analysts\Instacart Basket Analysis'

In [3]:
path

'D:\\CareerFoundary\\Immersion into Data Analystics\\4 Python Fundamentals for Data Analysts\\Instacart Basket Analysis'

In [4]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [5]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [6]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [8]:
df_ords.head()

Unnamed: 0  order_id  customer_id  order_number  orders_day_of_week  \
0           0   2539329            1             1                   2   
1           1   2398795            1             2                   3   
2           2    473747            1             3                   3   
3           3   2254736            1             4                   4   
4           4    431534            1             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

# 03. Task Practice Follow Along

In [9]:
# Create a dataframe
df_test = pd.DataFrame()

In [10]:
df_test['mix'] = ['a', 'b', 1, True]

In [11]:
display(df_test)

mix
0     a
1     b
2     1
3  True

In [12]:
# Check for mixed types
for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis =1)
    if len (df_test[weird]) > 0:
        print(col)

mix


In [13]:
df_test['mix'] = df_test['mix'].astype('str')

In [14]:
df_test.dtypes

mix    object
dtype: object

In [15]:
# Check for missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [16]:
# display missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [17]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [18]:
df_prods.shape

(49693, 5)

In [19]:
# remove missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [20]:
df_prods_clean.shape

(49677, 5)

In [21]:
# check for duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [22]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [23]:
# addressing duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [24]:
df_prods_clean_no_dups.shape

(49672, 5)

In [25]:
# exporting clean data
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

# 04. Task 4.5

## Q2. Run the df.describe() function on you df_prods dataframe. Share, in a markdown cell, whether anything about the data looks off or should be investigated further.

In [26]:
# running descriptives
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [32]:
df_prods[df_prods.prices == 99999]

product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0

### Answer for 2: The maximum for the product_id is lower than the count. This may be due to mising values or duplicates which should be addressed by the df_prods_clean_no_dups dataframe. More concerning, the maximum value for the prices is 99,999. Looking up the products associated with this price, it appears to be associated with 2% milk. This is either an error or a unclear place holder.


## Q3. Check for mixed-type data in your df_ords dataframe

In [30]:
# check for mix types

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [31]:
df_ords.dtypes

Unnamed: 0                  int64
order_id                    int64
customer_id                 int64
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

## Q4. If you find mixed-type data, fix it. 
### Answer for 4: There does not appear to be mixed data in the df_ords dataframe.


## Q5. Run a check for missing values in your df_ords data frame. 

In [33]:
# check for missing values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
customer_id                    0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [44]:
df_ords.value_counts(dropna = False)

Unnamed: 0  order_id  customer_id  order_number  orders_day_of_week  order_hour_of_day  days_since_prior_order
0           2539329   1            1             2                   8                  NaN                       1
2280791     1591157   137366       13            3                   15                 30.0                      1
2280715     1354759   137363       2             1                   10                 7.0                       1
2280716     1971373   137363       3             1                   10                 7.0                       1
2280717     1558866   137363       4             1                   10                 7.0                       1
                                                                                                                 ..
1140364     3266950   68521        12            0                   19                 20.0                      1
1140365     118963    68521        13            4                   19      

### Answer for 5: There are 206209 missing values from the days_since_prior_order columns. In exercise 4.4, when trying the value_counts () function on this column there were 206209 counts of NaN. As seen in the value counts above, 206209 is also the the number of customer IDs. Since this number represents the days since their prior order, the NaN likely indicates the customers first order; hence, there not being an available number and also matching the number of customer IDs. 


## Q6. Address the missing values using an appropriate method. In a markdown cell, explain why you used your method choice.

In [54]:
# create new column for new customer
df_ords_new['new_customer'] = df_ords['days_since_prior_order'].isnull() == True

In [53]:
df_ords_new.head(15)

Unnamed: 0  order_id  customer_id  order_number  orders_day_of_week  \
0            0   2539329            1             1                   2   
1            1   2398795            1             2                   3   
2            2    473747            1             3                   3   
3            3   2254736            1             4                   4   
4            4    431534            1             5                   4   
5            5   3367565            1             6                   2   
6            6    550135            1             7                   1   
7            7   3108588            1             8                   1   
8            8   2295261            1             9                   1   
9            9   2550362            1            10                   4   
10          10   1187899            1            11                   4   
11          11   2168274            2             1                   2   
12          12   1501582            2             2                   5   
13          13   1901567            2             3                   1   
14          14    738281            2             4                   2   

    order_hour_of_day  days_since_prior_order  new_customer  
0                   8                     NaN          True  
1                   7                    15.0         False  
2                  12                    21.0         False  
3                   7                    29.0         False  
4                  15                    28.0         False  
5                   7                    19.0         False  
6                   9                    20.0         False  
7                  14                    14.0         False  
8                  16                     0.0         False  
9                   8                    30.0         False  
10                  8                    14.0         False  
11                 11                     NaN          True  
12                 10                    10.0         False  
13                 10                     3.0         False  
14                 10                     8.0         False

### Answer for 6: In this case the 'missing values' actually give important information; this is the customers first order. We can tell from this who is a new customer or a returning customer. As suggested in the exercise, one way of dealing with this is to 'create a new variable that acts like a flag based on the missing value.' Which is what I did here; I created a new variable 'new_customer' based on the null values (where null = true [new customer]). 


## Q7. Run a check for duplicate values in your df_ords data. In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

In [56]:
# checking for duplicates
df_ords_dups = df_ords_new[df_ords_new.duplicated()]

In [57]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, customer_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, new_customer]
Index: []

### Answer for 7: no duplicates for df_ords_new 


## Q8. Address the duplicates using an appropriate method. In a markdown cell, explain why you used your method of choice.

### 8. Answer for 8: No duplicates found so did not have to be addressed.

## Q9. export your final, cleaned df_prods and df_ords data. as '.csv' files in your 'Prepared Data' folder and give them appropriate, succint names.

In [58]:
# exporting data 
df_ords_new.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_newcust.csv'))

### df_prods was already exported in practice exercise